django/db/migrations/executor.py

```diff
@@ -71,46 +72,91 @@ def migrate(self, targets, plan=None, fake=False, fake_initial=False):
         """
         if plan is None:
             plan = self.migration_plan(targets)
-        migrations_to_run = {m[0] for m in plan}
         # Create the forwards plan Django would follow on an empty database
         full_plan = self.migration_plan(self.loader.graph.leaf_nodes(), clean_start=True)
-        # Holds all states right before a migration is applied
-        # if the migration is being run.
+
+        all_forwards = all(not backwards for mig, backwards in plan)
+        all_backwards = all(backwards for mig, backwards in plan)
+
+        if all_forwards and not all_backwards:
+            self._migrate_all_forwards(plan, full_plan, fake=fake, fake_initial=fake_initial)
+        elif not all_forwards and all_backwards:
+            self._migrate_all_backwards(plan, full_plan, fake=fake)
+        elif plan:
```

Comment:
If `plan` is not passed to the method Django falls back on the `targets` and constructs the plan from them. Even if targets are given the plan might be empty (`[]`). This is a perfectly valid case (the database is in the state the user wants it to be).

I believe I could change the code to this. Does this make things more clear?
```python
        all_forwards = all(not backwards for mig, backwards in plan)
        all_backwards = all(backwards for mig, backwards in plan)

        if not plan:
            pass  # Nothing to do for an empty plan
        elif all_forwards == all_backwards:
            # This may only happen if there's a mixed plan or an error in the matrix
            raise InvalidMigrationPlan(...)
        elif all_forwards:
            self._migrate_all_forwards(...)
        elif all_backwards:
            # We could use `else` here because the if clause will always evaluate to true
            self._migrate_all_backwards(...)
```